iocore/net/QUICNetVConnection.cc

```diff
@@ -0,0 +1,645 @@
+    this->_packet_handler->send_packet(*packet, this);
+    this->_loss_detector->on_packet_sent(std::unique_ptr<const QUICPacket>(packet));
+  }
+
+  net_activity(this, this_ethread());
+
+  return QUICError(QUICErrorClass::NONE);
+}
+
+void
+QUICNetVConnection::_packetize_frames()
+{
+  uint32_t max_size = this->maximum_quic_packet_size();
+  uint32_t min_size = this->minimum_quic_packet_size();
+  ats_unique_buf buf(nullptr, [](void *p) { ats_free(p); });
+  size_t len = 0;
+
+  // Put frames into buf as many as possible
+  std::unique_ptr<QUICFrame, QUICFrameDeleterFunc> frame(nullptr, nullptr);
+  bool retransmittable                = false;
+  QUICPacketType previous_packet_type = QUICPacketType::UNINITIALIZED;
+  QUICPacketType current_packet_type  = QUICPacketType::UNINITIALIZED;
+
+  while (this->_frame_buffer.size() > 0) {
+    frame = std::move(this->_frame_buffer.front());
+    this->_frame_buffer.pop();
+    QUICRetransmissionFrame *rf = dynamic_cast<QUICRetransmissionFrame *>(frame.get());
+    previous_packet_type        = current_packet_type;
+    if (rf) {
+      current_packet_type = rf->packet_type();
+    } else {
+      current_packet_type = QUICPacketType::UNINITIALIZED;
+    }
+    if (len + frame->size() + MAX_PACKET_OVERHEAD > max_size || (previous_packet_type != current_packet_type && len > 0)) {
+      ink_assert(len > 0);
+      SCOPED_MUTEX_LOCK(transmitter_lock, this->get_transmitter_mutex().get(), this_ethread());
+      this->transmit_packet(this->_build_packet(std::move(buf), len, retransmittable, previous_packet_type));
+      len = 0;
+    }
+    retransmittable = retransmittable || (frame->type() != QUICFrameType::ACK && frame->type() != QUICFrameType::PADDING);
+
+    if (buf == nullptr) {
+      buf = ats_unique_malloc(max_size);
+    }
+    size_t l = 0;
+    DebugQUICCon("type=%s", QUICDebugNames::frame_type(frame->type()));
+    frame->store(buf.get() + len, &l);
+    len += l;
+  }
+
+  if (len != 0) {
+    // Pad with PADDING frames
+    if (min_size > len) {
+      // FIXME QUICNetVConnection should not know the actual type value of PADDING frame
+      memset(buf.get() + len, 0, min_size - len);
+      len += min_size - len;
+    }
+    SCOPED_MUTEX_LOCK(transmitter_lock, this->get_transmitter_mutex().get(), this_ethread());
+    this->transmit_packet(this->_build_packet(std::move(buf), len, retransmittable, current_packet_type));
+  }
+}
+
+void
+QUICNetVConnection::_recv_and_ack(const uint8_t *payload, uint16_t size, QUICPacketNumber packet_num)
+{
+  bool should_send_ack = this->_frame_dispatcher->receive_frames(payload, size);
+  this->_ack_frame_creator.update(packet_num, should_send_ack);
+  std::unique_ptr<QUICFrame, QUICFrameDeleterFunc> ack_frame = this->_ack_frame_creator.create_if_needed();
+  if (ack_frame != nullptr) {
+    this->transmit_frame(std::move(ack_frame));
+    eventProcessor.schedule_imm(this, ET_CALL, QUIC_EVENT_PACKET_WRITE_READY, nullptr);
```

Comment:
I guess `schedule_imm` in `transmit_frame` was added for convenience so we don't need to call `schedule_imm` after calling `transmit_frame`.

However, scheduling an event on every frame is too expensive, so I think we should add `flush()`, or `bool transmit_now` parameter.